runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6937,16 +6937,57 @@ J9::Z::TreeEvaluator::VMifInstanceOfEvaluator(TR::Node * node, TR::CodeGenerator
    return NULL;
    }
 
+TR::Register *
+J9::Z::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::Register *objReg, TR::LabelSymbol* mergeLabel, TR::LabelSymbol *helperCallLabel, TR::CodeGenerator *cg)
+   {
+   TR::Register *objectClassReg = cg->allocateRegister();
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, objectClassReg, generateS390MemoryReference(objReg, TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+   TR::Register *classFlagsRegister = cg->allocateRegister();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->comp()->fe());
+   generateRXInstruction(cg, TR::InstOpCode::L, node, classFlagsRegister, generateS390MemoryReference(objectClassReg, static_cast<uint32_t>(fej9->getOffsetOfClassFlags()), cg));
+   bool generateOOLSection = helperCallLabel == NULL;
+   if (generateOOLSection)
+      helperCallLabel = generateLabelSymbol(cg);
+   generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, classFlagsRegister, static_cast<int32_t>(J9ClassIsValueType), TR::InstOpCode::COND_BE, helperCallLabel);
```

Comment:
I don't think we need to load the flags into a register. Can we get away with using `TM` to just test in memory and generate a `BRC` to act on it?

Reply:
I took a look at the `TM`, unfortunately we can not use that instruction here as in the instruction, immediate field is only 8-bit wide, while the constant `J9ClassIsValueType` takes up 12 bits [1]. Even we will not be using the Compare and Branch instruction which compares register with immediate and branches.
So currently what it will generate looks like following. 

```
L classFlagsRegister,@(objectClassReg+OffsetOfClassFlags)
CFI classFlagsRegister,J9ClassIsValueType
BRC COND_BE, helperCallLabel
```
[1] https://github.com/eclipse/openj9/blob/7d740b427a60bb8808aa5ec959d98d002d21c8d1/runtime/oti/j9nonbuilder.h#L76

Reply:
Actually I interpreted x86 implementation wrong. A version of `TM` should be used here not logical compare. @fjeremic  marking this as WIP while I fix it.

Reply:
Correct sequence will look like following. 
```
L classFlagsRegister,@(J9class+classFlags)
TMLL classFlagsRegister,0x400 // Ands 48-63 bits of classFlagsRegister with 0x400
BRC NotZero, helperCallLabel
```
Changed should be ok, though marking this WIP till I test it out again.

Reply:
Yeah that sequence looks good.